File: apps/ios/Packages/module-ai/Sources/ModuleAI/Application/UseCases/AIChat/SendFunctionMessageUseCase.swift
//
//  SendFunctionMessageUseCase.swift
//  ModuleAI
//
//  Created by Codex on 2026/1/8.
//

import Foundation

struct SendFunctionMessageUseCase {
    private let repository: AIChatRepository

    init(repository: AIChatRepository) {
        self.repository = repository
    }

    func execute(
        text: String,
        historyId: Int?
    ) async throws -> UseCaseResult<SendFunctionMessageOutput> {
        let model = try await repository.sendFunctionMessage(text, historyId: historyId)

        guard let historyId = model.historyId else {
            return .failure(.message(model.msg))
        }

        guard let hasTool = model.hasTool,
              hasTool,
              let name = model.name,
              let arguments = model.arguments else {
            return .failure(.message(model.msg))
        }

        if let intentType = AIChatIntentResolver.resolve(arguments: arguments) {
            return .success(.intent(text: text, type: intentType))
        }

        return .success(.chartRequest(name: name, historyId: historyId, arguments: arguments))
    }
}

File: app-android/feature/ai-chat/src/main/java/com/aidatainsight/android/feature/aichat/application/usecase/SendFunctionMessageUseCase.kt
package com.aidatainsight.android.feature.aichat.application.usecase

import com.aidatainsight.android.feature.aichat.application.AIChatIntentResolver
import com.aidatainsight.android.feature.aichat.application.model.SendFunctionMessageOutput
import com.aidatainsight.android.feature.aichat.application.model.UseCaseResult
import com.aidatainsight.android.feature.aichat.domain.AIChatRepository

class SendFunctionMessageUseCase(
    private val repository: AIChatRepository,
) {
    suspend operator fun invoke(
        text: String,
        historyId: Int?,
    ): UseCaseResult<SendFunctionMessageOutput> {
        val model = repository.sendFunctionMessage(text, historyId)
        val nextHistoryId = model.historyId
            ?: return UseCaseResult.Failure(model.msg)

        val name = model.name
        val arguments = model.arguments
        if (model.hasTool != true || name == null || arguments == null) {
            return UseCaseResult.Failure(model.msg)
        }

        val intentType = AIChatIntentResolver.resolve(arguments)
        if (intentType != null) {
            return UseCaseResult.Success(
                SendFunctionMessageOutput.Intent(text = text, type = intentType),
            )
        }

        return UseCaseResult.Success(
            SendFunctionMessageOutput.ChartRequest(
                name = name,
                historyId = nextHistoryId,
                arguments = arguments,
            ),
        )
    }
}
